### PR TITLE
fix(server-functions): omitted any from fieldErrors

### LINKED ISSUE #6719: [🐞] fieldErrors - Type instantiation is excessively deep and possibly infinite.
### Which component is affected?

Qwik City (routing)

### Describe the bug

I implemented this feature myself, turns out my type is too complex for typescript and not sure how to fix it.
My plan is to check this out later, I think it's connected with `any` in `zod`.

`routeAction$` with complex zod ends up in the typescript message:

> Type instantiation is excessively deep and possibly infinite.

See https://stackblitz.com/edit/qwik-starter-ps6npq?file=src%2Froutes%2Findex.tsx

### Reproduction

https://stackblitz.com/edit/qwik-starter-ps6npq?file=src%2Froutes%2Findex.tsx

### Steps to reproduce

_No response_

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.20.3 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    @builder.io/qwik: ^1.7.1 => 1.7.1 
    @builder.io/qwik-city: ^1.7.1 => 1.7.1 
    typescript: 5.3.3 => 5.3.3 
    undici: 5.28.2 => 5.28.2 
    vite: 4.5.2 => 4.5.2
```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by tzdesign**
Typescript crashes because any could be infinite

fix #6719

# What is it?

- [ ] Feature / enhancement
- [x] Bug
- [ ] Docs / tests / types / typos

# Description

See

- #6719 

# Checklist:

- [x] My code follows the [developer guidelines of this project](https://github.com/QwikDev/qwik/blob/main/CONTRIBUTING.md)
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] Added new tests to cover the fix / functionality

### CODE REVIEW

**Comment by tzdesign** [comment]
@gioboa @shairez Looks like I've tripped myself up. We sometimes use any because we have multi-step forms. Typescript unfortunately can't handle it. Showing: `Type instantiation is excessively deep and possibly infinite.`

I rewrote it so that the type never ends up in fieldErrors because Zod ignores any anyway.

I hope one of you can write typescript a little better than I can. I have patched it with the small change to
```{} extends T[K]``` which excludes all any types. Let's see if it still works like this in later Typescript versions.

File: starters/apps/qwikcity-test/src/routes/(common)/actions/issue5463/index.tsx [hunk 1]
```diff
@@ -19,6 +19,7 @@ export const useDotNotationAction = globalAction$(
           firstName: z.string(),
         }),
       }),
+      someAny: z.any(), // test for making sure it does not add it to the fieldErrors see #6719
```
Comment (wmertens) [inline starters/apps/qwikcity-test/src/routes/(common)/actions/issue5463/index.tsx:22]: I think we need an explicit test for it in the spec file

File: packages/qwik-city/runtime/src/types.ts [hunk 2]
```diff
@@ -364,13 +364,15 @@ export type FailOfRest<REST extends readonly DataValidator[]> = REST extends rea
...
 export type ValidatorErrorKeyDotNotation<T, Prefix extends string = ''> = T extends object
   ? {
-      [K in keyof T & string]: T[K] extends (infer U)[]
-        ? U extends object
-          ? `${Prefix}${K}[]` | `${Prefix}${K}[]${ValidatorErrorKeyDotNotation<U, '.'>}`
-          : `${Prefix}${K}[]`
-        : T[K] extends object
-          ? ValidatorErrorKeyDotNotation<T[K], `${Prefix}${K}.`>
-          : `${Prefix}${K}`;
+      [K in keyof T & string]: {} extends T[K]
```
Comment (wmertens) [inline packages/qwik-city/runtime/src/types.ts:367]: Can you test this type? You can find examples of type tests in the `packages/qwik/src/core/render/jsx/types/jsx-types.unit.tsx` for example.

**Review by wmertens — CHANGES_REQUESTED** [review]

File: packages/qwik-city/runtime/src/types.ts (hunk 2 above)
Comment (tzdesign) [reply packages/qwik-city/runtime/src/types.ts:367]: I will check it out after my short vacation on Thursday.

I think this change will remove the dot notation, but let me test everything. 

The idea of testing the type is perfect, will do it too.

File: starters/apps/qwikcity-test/src/routes/(common)/actions/issue5463/index.tsx (hunk 1 above)
Comment (tzdesign) [reply starters/apps/qwikcity-test/src/routes/(common)/actions/issue5463/index.tsx]: @wmertens added a test in `server-functions.unit.ts`.

But `expectTypeOf` is pretty weird. I found some issues in vitest and expect-type implementation:
See https://github.com/mmkal/expect-type/issues/31

The any implementation with not seems to work. But I don't know if this will fail at least on build.full.

Let me know what you think.

**Comment by tzdesign** [comment]
@fabian-hiller just to let you know the ValidatorErrorKeyDotNotation wasn't capturing Record types. Added IsAny to the types.ts in the runtime and added tests.

Maybe you can catch up with me some time talking about rewriting this a bit more.
There are still edge-cases for arrays in my concept, but it's better than having the top-level keys with flatten rod errors.

**Comment by fabian-hiller** [comment]
Have you seen my PR #6752?

**Comment by wmertens** [comment]
@fabian-hiller I'd like to merge this already, it will cause your pr to get merge conflicts though. Ok?

File: packages/qwik-city/src/runtime/src/types.ts [hunk 3]
```diff
@@ -362,17 +362,27 @@ export type FailOfRest<REST extends readonly DataValidator[]> = REST extends rea
...
-      [K in keyof T & string]: T[K] extends (infer U)[]
-        ? U extends object
-          ? `${Prefix}${K}[]` | `${Prefix}${K}[]${ValidatorErrorKeyDotNotation<U, '.'>}`
-          : `${Prefix}${K}[]`
-        : T[K] extends object
-          ? ValidatorErrorKeyDotNotation<T[K], `${Prefix}${K}.`>
-          : `${Prefix}${K}`;
-    }[keyof T & string]
-  : never;
+export type IsAny<T> = 0 extends T & 1 ? true : false;
```
Comment (fabian-hiller) [inline packages/qwik-city/src/runtime/src/types.ts:365]: I think this is just a util type similar to the `Prettify` type above and should not be exported.

**Comment by fabian-hiller** [comment]
@wmertens and @tzdesign this PR contains the same `any` fix that #6752. To avoid merge conflicts it could be simpler to just move over the `server-functions.unit.ts` file to my PR and add Tobi as a co-author.

**Comment by tzdesign** [comment]
@fabian-hiller I'm fine with it, but I copied your type and your type includes any as type string.  So I added a bit more logic.

I'm fine merging the prs together.

**Comment by fabian-hiller** [comment]
> I copied your type and your type includes any as type string. So I added a bit more logic.

Can you explain this? Maybe I missed something.

**Comment by fabian-hiller** [comment]
Ahh... So you recommend that I also copy your entire `ValidatorErrorKeyDotNotation` type?

**Comment by wmertens** [comment]
> Ahh... So you recommend that I should also copy your entire `ValidatorErrorKeyDotNotation` type?

I think that's a good idea

**Comment by fabian-hiller** [comment]
Done

**Comment by tzdesign** [comment]
@fabian-hiller perfect!

Thank you guys.
